crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -449,6 +470,116 @@ impl<'db> TypeInferenceBuilder<'db> {
         self.infer_expression(expression.node_ref(self.db));
     }
 
+    fn invalid_assignment_diagnostic(
+        &mut self,
+        node: AnyNodeRef,
+        declared_ty: Type<'db>,
+        assigned_ty: Type<'db>,
+    ) {
+        match declared_ty {
+            Type::Class(class) => {
+                self.add_diagnostic(node, "invalid-assignment", format_args!(
+                        "Implicit shadowing of class '{}'; annotate to make it explicit if this is intentional.",
+                        class.name(self.db)));
+            }
+            Type::Function(function) => {
+                self.add_diagnostic(node, "invalid-assignment", format_args!(
+                        "Implicit shadowing of function '{}'; annotate to make it explicit if this is intentional.",
+                        function.name(self.db)));
+            }
+            _ => {
+                self.add_diagnostic(
+                    node,
+                    "invalid-assignment",
+                    format_args!(
+                        "Object of type '{}' is not assignable to '{}'.",
```

Comment:
Nit: Maybe just use the term `Value` instead of `Object` because not all users know that all python values inherit from object.

Reply:
I prefer "object", because it's consistent with the terminology users see all the time in Python's tracebacks at runtime. I don't see it as referring to the fact that all Python values inherit from `builtins.object`; I think it's referring to the more fundamental semantic that all values in Python are, conceptually, objects